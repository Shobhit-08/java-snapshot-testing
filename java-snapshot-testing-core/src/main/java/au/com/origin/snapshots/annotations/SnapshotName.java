package au.com.origin.snapshots.annotations;

import java.lang.annotation.*;

@Documented
@Target({ElementType.METHOD})
@Inherited
@Retention(RetentionPolicy.RUNTIME)
public @interface SnapshotName {
  String value();
}
