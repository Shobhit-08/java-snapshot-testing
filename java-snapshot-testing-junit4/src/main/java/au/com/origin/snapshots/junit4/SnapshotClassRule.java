package au.com.origin.snapshots.junit4;

import au.com.origin.snapshots.SnapshotVerifier;
import lombok.Getter;
import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

public class SnapshotClassRule implements TestRule {

  @Getter private SnapshotVerifier snapshotVerifier;

  @Getter private SharedSnapshotHelpers helpers = new SharedSnapshotHelpers();

  @Override
  public Statement apply(Statement base, Description description) {
    return new Statement() {
      @Override
      public void evaluate() throws Throwable {
        snapshotVerifier = helpers.getSnapshotVerifier(description);
        try {
          base.evaluate();
        } finally {
          snapshotVerifier.validateSnapshots();
        }
      }
    };
  }
}
