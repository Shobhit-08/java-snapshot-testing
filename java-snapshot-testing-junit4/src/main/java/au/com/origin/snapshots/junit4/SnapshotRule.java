package au.com.origin.snapshots.junit4;

import au.com.origin.snapshots.exceptions.SnapshotExtensionException;
import lombok.RequiredArgsConstructor;
import org.junit.rules.MethodRule;
import org.junit.runners.model.FrameworkMethod;
import org.junit.runners.model.Statement;

@RequiredArgsConstructor
public class SnapshotRule implements MethodRule {

  private final SnapshotClassRule snapshotClassRule;

  @Override
  public Statement apply(Statement base, FrameworkMethod method, Object test) {
    final SharedSnapshotHelpers helpers = snapshotClassRule.getHelpers();
    helpers.injectExpectInstanceVariable(
        snapshotClassRule.getSnapshotVerifier(), method.getMethod(), test);
    if (helpers.hasExpectArgument(method)) {
      throw new SnapshotExtensionException(
          "Sorry, we don't support 'Expect' as a method argument for @Rule or @ClassRule. "
              + "Please use an instance variable or @RunWith(SnapshotRunner.class) instead.");
    }

    return new Statement() {
      @Override
      public void evaluate() throws Throwable {
        base.evaluate();
      }
    };
  }
}
