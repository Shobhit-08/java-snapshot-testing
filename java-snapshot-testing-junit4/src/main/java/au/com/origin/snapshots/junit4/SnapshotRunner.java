package au.com.origin.snapshots.junit4;

import au.com.origin.snapshots.SnapshotVerifier;
import au.com.origin.snapshots.logging.LoggingHelper;
import java.util.List;
import lombok.extern.slf4j.Slf4j;
import org.junit.Test;
import org.junit.runner.notification.RunNotifier;
import org.junit.runners.BlockJUnit4ClassRunner;
import org.junit.runners.model.FrameworkMethod;
import org.junit.runners.model.InitializationError;
import org.junit.runners.model.Statement;

/**
 * Runner to enable java-snapshot-testing
 *
 * <p>If you are already using @RunWith for something else such as @RunWith(Parameterized.class) use
 * these Rules instead.
 *
 * @see SnapshotClassRule
 * @see SnapshotRule
 *     <pre>{@code
 * {@literal @}ClassRule
 * public static SnapshotClassRule snapshotClassRule = new SnapshotClassRule();
 *
 * {@literal @}Rule
 * public SnapshotRule snapshotRule = new SnapshotRule(snapshotClassRule);
 *
 * private Expect expect;
 * }</pre>
 *     Loosely based on:
 *     https://stackoverflow.com/questions/27745691/how-to-combine-runwith-with-runwithparameterized-class
 */
@Slf4j
public class SnapshotRunner extends BlockJUnit4ClassRunner {

  SnapshotVerifier snapshotVerifier;

  private SharedSnapshotHelpers helpers = new SharedSnapshotHelpers();

  public SnapshotRunner(Class<?> klass) throws InitializationError {
    super(klass);
  }

  @Override
  protected Statement methodInvoker(FrameworkMethod method, Object test) {
    boolean isTest = method.getMethod().isAnnotationPresent(Test.class);
    if (isTest) {
      helpers.injectExpectInstanceVariable(snapshotVerifier, method.getMethod(), test);
      boolean shouldInjectMethodArgument = helpers.hasExpectArgument(method);
      if (shouldInjectMethodArgument) {
        LoggingHelper.deprecatedV5(
            log,
            "Injecting 'Expect' via method a argument is no longer recommended. Consider using instance variable injection instead.");
        return helpers.injectExpectMethodArgument(snapshotVerifier, method, test);
      }
    }

    return super.methodInvoker(method, test);
  }

  @Override
  public void run(RunNotifier notifier) {
    snapshotVerifier = helpers.getSnapshotVerifier(getDescription());
    super.run(notifier);
    snapshotVerifier.validateSnapshots();
  }

  @Override
  protected void validateTestMethods(List<Throwable> errors) {
    // Disable as it checks for zero arguments
  }
}
