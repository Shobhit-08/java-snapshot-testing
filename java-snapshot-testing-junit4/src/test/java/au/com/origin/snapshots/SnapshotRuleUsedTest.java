package au.com.origin.snapshots;

import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit4.SnapshotClassRule;
import au.com.origin.snapshots.junit4.SnapshotRule;
import org.junit.ClassRule;
import org.junit.Rule;
import org.junit.Test;

public class SnapshotRuleUsedTest {

  @ClassRule public static SnapshotClassRule snapshotClassRule = new SnapshotClassRule();

  @Rule public SnapshotRule snapshotRule = new SnapshotRule(snapshotClassRule);

  private Expect expect;

  @Test
  public void shouldUseExtensionViaInstanceVariable() {
    this.expect.toMatchSnapshot("Hello World");
  }

  @Test
  public void shouldUseExtensionAgainViaInstanceVariable() {
    this.expect.toMatchSnapshot("Hello World");
  }

  @SnapshotName("hello_world")
  @Test
  public void shouldUseExtensionWithSnapshotName() {
    expect.toMatchSnapshot("Hello World");
  }
}
