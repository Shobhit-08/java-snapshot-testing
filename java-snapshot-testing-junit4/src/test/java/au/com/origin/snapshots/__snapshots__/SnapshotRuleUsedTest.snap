au.com.origin.snapshots.SnapshotRuleUsedTest.shouldUseExtensionAgainViaInstanceVariable=[
Hello World
]


au.com.origin.snapshots.SnapshotRuleUsedTest.shouldUseExtensionViaInstanceVariable=[
Hello World
]


hello_world=[
Hello World
]